Marking Refactor, Adds tail markings + Alt heads + Secondary (F)Hair themes, Darkens Vulp/Taj + Unathi horns/frills, Sprite adjustments and more...

KasparoVy opened the pull request:
**This is the re-up.** The SQL has been changed to use varchar(255) instead of mediumtext for marking colours and styles, which means the default value parts of the 'required SQL changes' query will work. I'm quite nervous about putting this back up, which was why the PR's return took a few days instead of one. I went back through everything, corrected a minor issue, cleared unneeded srcs in appearance.dm as per TheDZD's suggestion and tested the SQL schema by creating a fresh database. _The SQL query in the 'required SQL changes' has been amended since the last PR,_ and has been tested on a fresh database built from the current live SQL schema populated with dummy characters to check for validity.

**_CREDITS:_**
Tail marking idea and Vox tail marking sprites by @Fullofskittles.
Most of the new body and head marking sprites by @Fullofskittles.
5 new Tajaran hair sprites from **Polaris(?)**.
Sharp Unathi snout inspired by **TG**.
color2R/G/B helper procs and PR overhaul suggested/co-authored by @Krausus.

## IMPORTANT NOTICE:
**Vulpkanin and Tajara body sprites have been darkened! Characters will have to adjust their body colours if they want to look the same.**
- _Add 27 to the R, G and B values if Vulpkanin._
- _Add 15 to the R, G and B values if Tajara._

**Tiger body markings have been darkened! Characters will have to adjust their body marking colours if they want to look the same.**
- _Add 14 to the R, G and B values if body/head marking has 'tiger' in it._

**Vulpkanin facial fur patterns have been darkened! Characters will have to adjust their head accessory/facial hair colours if they want to look the same.**
- _Add 23 to the R, G and B values of your facial hair/head accessories._

**Unathi frills and horns have been darkened! Characters will have to adjust their facial hair colours if they want to look the same.**
- _Add 27 to the R, G and B values of your frills. You're going to have to do the webbing colour from scratch._
- _Add 21 to the R, G and B values of your horns._

**Tajaran head/facial hair has been darkened! Characters will have to adjust their head/facial hair colours if they want to look the same.**
- _Add 28 to the R, G and B values of your head/facial hair._

**Vulpkanin head/facial hair has been darkened! Characters will have to adjust their head hair colours if they want to look the same.**

For Head Hair:
- _If Adhara, add 12 to R, G and B values for hair colour._
- _If Anita, add 10 to R, G and B values for hair colour._
- _If Apollo, add 6 to R, G and B values for hair colour._
- _If Belle, add 6 to R, G and B values for hair colour._
- _If Bun, add 5 to R, G and B values for hair colour._
- _If Curl, add 15 to R, G and B values for hair colour._
- _If Hawk, don't do anything._
- _If Jagged, add 10 to R, G and B values for hair colour._
- _If Kajam, add 16 to R, G and B values for hair colour._
- _If Keid, add 27 to R, G and B values for hair colour._
- _If Kleeia, add 14 to R, G and B values for hair colour._
- _If Mizar, add 14 to R, G and B values for hair colour._
- _If Short, add 7 to R, G and B values for hair colour._
- _If Spike, add 6 to R, G and B values for hair colour._

For Facial Hair:
- _If Kita, add 42 to R, G and B values for facial hair colour._
- _If anything else, don't do anything._

## Features, fixes and changes:
:cl:
tweak: Refactors markings. Now split into head, body and tail markings.
tweak: Refactors morph again. Changes made while morphing are now reflected instantly on your sprite.
tweak: Darkens Vulpkanin and Tajara for improved colour fidelity. Species base colour changed to compensate. Add 27 to R, G and B values if Vulp, 15 if Taj to keep the same colour on your character.
tweak: Darkens Vulpkanin facial fur patterns for improved colour fidelity. Add 23 to the R, G and B values of your facial hair/head accessory to keep the same colour on your character.
tweak: Darkens Unathi frills and horns. Add 27 to the R, G and B values of your frills and 21 for horns to keep the same colour on your character.
tweak: Vulpkanin tails are now split-rendered. Tails overlap hair when facing north, but are overlapped by hair when facing all other directions. No other differences, only affects the fluffy husky tail.
tweak: The heads of tiger markings have been isolated and can now be coloured independently.
tweak: Hair can now cover facial hair, and glasses are no longer covered by facial hair and head accessories. Means that two-piece Vulpkanin facial fur patterns won't almost entirely cover whatever glasses they're wearing.
tweak: Tajaran head and facial hair have been darkened. Add 28 to the R, G and B values of your Tajara head/facial hair to keep the same colour on your character.
tweak: Vulpkanin head hair and one facial hair style has been darkened. Refer to PR notes for darkening factors.
tweak: The HAS_MARKINGS flag now represents all marking locations. The HAS_location_MARKINGS flags are for specific locations (head, body, tail).
tweak: Improves the quality of some Vulpkanin/Tajara/Unathi mask sprites.
fix: Naked Humans and Skrell will no longer have exposed flesh poking out their wrists when facing east or west. Male Humans now no longer lack a butt pixel when facing east or west.
fix: Mutated limbs now render correctly as soon as they're mutated.
fix: Mobs will now correctly appear as fat/skinny.
fix: Species with TAIL_OVERLAPPED will no longer wag tail at max speed when facing north.
fix: Species with TAIL_OVERLAPPED and body accessories won't crash the game when they wag.
fix: Tiger Head and Face markings adjusted so they don't look ugly on certain species' ears.
fix: Fixes issue where Unathi dorsal stripe didn't render.
fix: Tajaran ears are no longer in their hair sprites, but on their actual heads.
fix: The adminbus body accessory now overrides the TAIL_OVERLAPPED flag and will now be rendered correctly on a body regardless of species bodyflags.
fix: Changing a person's species via CMA will no longer result in a mob with all the same cosmetic attributes as the old species.
fix: Dressers now have species fitting checks for undergarments. Drask are now able to wear undershirts and underwear like they were originally meant to.
rscadd: Adds some head markings for Tajara and Vulpkanin.
rscadd: Adds body marking(s) for Tajara, Vulpkanin, Unathi and Drask.
rscadd: Adds tail markings for Vox and Vulpkanin.
rscadd: Adds 5 Tajaran hairstyles.
rscadd: Adds a tweaked version of an existing Vox hairstyle. Shouldn't clip with certain jackets as much.
rscadd: Adds a hairstyle for Humans, Unathi and Vulpkanin.
rscadd: Adds a facial hairstyle for Vulpkanin.
rscadd: Adds secondary hair (head and facial) themes. Skrell can now colour their tentacle cloths independently, the beads in the Tajaran/Hippie braid hairstyles can be coloured and the same with the webbing in Unathi frills. Can be changed at nanomirrors, via morph and CMA.
rscadd: Adds alternate heads. Unathi can now choose a head with a sharp snout. Can be changed via morph and CMA.
rscadd: Helper procs to convert a hex colour into either R, G, or B.
rscadd: Markings, body accessories, head accessories, alt heads and their colours (if they have colours) can now be randomized at the character preferences screen.
rscadd: Adds a system that attempts to fit mobs with alternate heads with appropriate mask sprites. The sprites just need to be suffixed with the alt_head suffix in sprite_accessories.dm.
imageadd: Adds mask sprites tweaked to fit the new Unathi alternate head to go along with the new alt. head mask fitting system.
/:cl:

## Required SQL Changes:
```
ALTER TABLE `characters` DROP COLUMN `marking_style_name`;
ALTER TABLE `characters` ADD COLUMN `marking_styles` varchar(255) NOT NULL DEFAULT 'head=None&body=None&tail=None' AFTER `facial_style_name`;
UPDATE `characters` SET `marking_styles` = 'head=None&body=None&tail=None';
ALTER TABLE `characters` DROP COLUMN `markings_red`, DROP COLUMN `markings_green`, DROP COLUMN `markings_blue`;
ALTER TABLE `characters` ADD COLUMN `marking_colours` varchar(255) NOT NULL DEFAULT 'head=%23000000&body=%23000000&tail=%23000000' AFTER `skin_blue`;
UPDATE `characters` SET `marking_colours` = 'head=%23000000&body=%23000000&tail=%23000000';
UPDATE `characters` SET `hair_style_name` = "Bald" WHERE `hair_style_name` = "Tajaran Ears";
ALTER TABLE `characters` ADD COLUMN `alt_head_name` varchar(45) NOT NULL AFTER `head_accessory_style_name`;
UPDATE `characters` SET `hair_style_name` = "Bald" WHERE `hair_style_name` = "None";
ALTER TABLE `characters` ADD COLUMN `secondary_facial_red` smallint(4) NOT NULL AFTER `facial_blue`;
ALTER TABLE `characters` ADD COLUMN `secondary_facial_green` smallint(4) NOT NULL AFTER `secondary_facial_red`;
ALTER TABLE `characters` ADD COLUMN `secondary_facial_blue` smallint(4) NOT NULL AFTER `secondary_facial_green`;
ALTER TABLE `characters` ADD COLUMN `secondary_hair_red` smallint(4) NOT NULL AFTER `hair_blue`;
ALTER TABLE `characters` ADD COLUMN `secondary_hair_green` smallint(4) NOT NULL AFTER `secondary_hair_red`;
ALTER TABLE `characters` ADD COLUMN `secondary_hair_blue` smallint(4) NOT NULL AFTER `secondary_hair_green`;
```

## More pictures coming.
_Vulpkanin Darkening_
![vulpdarkening](https://cloud.githubusercontent.com/assets/12377767/16969884/2e031cdc-4de6-11e6-869b-df661182a518.png)

_The reason why Vulpkanin tails are now split-rendered._
![vulptailsplit](https://cloud.githubusercontent.com/assets/12377767/16969893/3c3ee95c-4de6-11e6-9e7e-ce6a6018281e.png)

_Vulpkanin chargen._
![markingsvulp](https://cloud.githubusercontent.com/assets/12377767/16974218/6638df66-4e08-11e6-8fa9-60bc9209d3aa.PNG)

_Unathi chargen. The 'colour # 2' options only pop up if the selected hair/facial hair style has secondary themes and can be coloured._
![unathichargen](https://cloud.githubusercontent.com/assets/12377767/16974246/adb6a242-4e08-11e6-846b-664a5694f42f.PNG)

_Unathi in-game feat. hair with secondary theme (beads) and facial hair with secondary theme (webbing)._
![unathiingameid](https://cloud.githubusercontent.com/assets/12377767/16974501/31c36fdc-4e0b-11e6-8bc6-6c0c7682f1cf.png)

_Vulpkanin in-game feat. hair with secondary theme (beads), chest crest and tail markings._
![vulpingame](https://cloud.githubusercontent.com/assets/12377767/16974560/b701e00c-4e0b-11e6-8e1d-f857b0a0d849.png)

_Vulpkanin Tail-marking anims working as intended._
![vulptailmarkingwag](https://cloud.githubusercontent.com/assets/12377767/16974905/87874606-4e0f-11e6-8767-836bef7b5dde.gif)

_Vox Tail-marking anims working as intended. No more top-speed wag when facing north._
![voxtailmarkingsstatic](https://cloud.githubusercontent.com/assets/12377767/16975014/847151ae-4e10-11e6-889d-5607fcf68b3e.png)![voxtailmarkingwag](https://cloud.githubusercontent.com/assets/12377767/16975015/8471cef4-4e10-11e6-8fb7-1c7aeb4e1e08.gif)![voxtailmarkingswag](https://cloud.githubusercontent.com/assets/12377767/16975016/8472234a-4e10-11e6-9bff-5757bde5f9ac.gif)

_Seeing that we're able to change tail markings in the admin CMA panel._
![cmatailmarkings](https://cloud.githubusercontent.com/assets/12377767/16975123/60870fb2-4e11-11e6-8dc1-5f55a4a36d55.PNG)

_Seeing that we're able to change LOTS of stuff in the admin CMA panel. Keep in mind that you can change secondary hair and secondary facial hair colours at a SalonPro Nanomirror(tm) and that you can change all this via Morph as well._
![unathiadmincma](https://cloud.githubusercontent.com/assets/12377767/16975128/6f8d6178-4e11-11e6-8c62-3334ea68e9ef.PNG)

_Skrell can now independently colour their tentacle cloths. Was going to be part of another PR but the idea fit in much better here._
![skrelltentaclesclothcolour](https://cloud.githubusercontent.com/assets/12377767/16976020/c1797762-4e19-11e6-86fe-e8f151b0eb5e.png)

_Darkened Tajara as well in the same manner as Vulpkanin._
![tajdarkening](https://cloud.githubusercontent.com/assets/12377767/16979162/9d74b1ba-4e2d-11e6-87ec-a1b887f7279b.png)

_Tattoos using Humans as an example species._
![tatsonhumans](https://cloud.githubusercontent.com/assets/12377767/17078758/51c898aa-50c9-11e6-900c-d7dc0569cc29.png)

_Hippie braid hairstyle (including the OG on Tajara. Also showcases some of the other new cosmetic features.)_
![hippiebraid](https://cloud.githubusercontent.com/assets/12377767/17079648/cef33d1a-50e4-11e6-92e9-fcfbb08bcfb1.png)

_The new Tajaran hair styles from Polaris._
_Top row: L - Curly, Mid - Ladies' Retro, R - Victory Curls_
![tajaranhair](https://cloud.githubusercontent.com/assets/12377767/17079649/e1d6113c-50e4-11e6-9041-9d0b6c2f98a8.png)
_Bottom row: L - Bob, Mid - Finger Curls_

_Foxes! Showcasing more of the markings._
![foxes](https://cloud.githubusercontent.com/assets/12377767/17081531/50ffc6fc-512b-11e6-8ca6-41c57859ea65.png)
![foxwag1](https://cloud.githubusercontent.com/assets/12377767/17081532/53e5fb16-512b-11e6-8e68-f3afee4bab69.gif)![foxwag2](https://cloud.githubusercontent.com/assets/12377767/17081533/53f2cb02-512b-11e6-824d-3d17d4c18a90.gif)

_Yet more of the new Vulpkanin markings._
![shibevulpmarkings](https://cloud.githubusercontent.com/assets/12377767/17888636/de61ad22-68f9-11e6-9fe5-de28ae79e901.png)![dobermanvulpmarkings](https://cloud.githubusercontent.com/assets/12377767/17888638/de69507c-68f9-11e6-877c-b0426c3051a1.png)

_The new Drask marking in action._
![draskspinemarking](https://cloud.githubusercontent.com/assets/12377767/17888609/bfcaaf9e-68f9-11e6-9813-a437a04e1977.png)

_The rest of the Vulpkanin tail markings._
![othertailmarkings](https://cloud.githubusercontent.com/assets/12377767/17081537/703d732a-512b-11e6-9037-8443a5f22e1f.png)
![silverfoxwag1](https://cloud.githubusercontent.com/assets/12377767/17081538/75890f4c-512b-11e6-9258-bbc712e229fb.gif)![silverfoxwag3](https://cloud.githubusercontent.com/assets/12377767/17081539/759a007c-512b-11e6-95ce-a512215e17fc.gif)![silverfoxwag2](https://cloud.githubusercontent.com/assets/12377767/17081540/75a02196-512b-11e6-844e-1a21387025ee.gif)

_Glasses are no longer covered by facial hair and head accessories. You can now see glasses on Vulpkanin._
![glassesfix](https://cloud.githubusercontent.com/assets/12377767/17238187/94607216-5527-11e6-9fb9-f81ce14318d5.png)

_Alt. head mask fitting system in action. The first Unathi pictured has the new alt. head._
Top from left to right: Unathi with sharp snout, Unathi with default snout.
![altheadmaskhandling](https://cloud.githubusercontent.com/assets/12377767/17735514/9e465358-6450-11e6-818a-dffbb1b0d547.png)
Bottom from left to right: Tajara, Human.

_Your body can now be fat/skinny (if your species and its sprites allow)._ ![bigboi](https://cloud.githubusercontent.com/assets/12377767/17889749/dc8530ae-68ff-11e6-9cd1-261b54fa2d94.PNG)

_Your body will now immediately show deformities when mutated._ ![ohgod](https://cloud.githubusercontent.com/assets/12377767/17892761/e5e3d87e-6910-11e6-92b8-ab4d55e757b5.png)

Resolves #5478

Krausus commented on code/game/dna/dna2_helpers.dm:
```diff
@@ -152,10 +152,10 @@
+		var/list/marking_colours = params2list(H.m_colours)
+		marking_colours["head"] = "#[num2hex(dna.GetUIValueRange(DNA_UI_HEAD_MARK_R,255),2)][num2hex(dna.GetUIValueRange(DNA_UI_HEAD_MARK_G,255),2)][num2hex(dna.GetUIValueRange(DNA_UI_HEAD_MARK_B,255),2)]"
+		marking_colours["body"] = "#[num2hex(dna.GetUIValueRange(DNA_UI_BODY_MARK_R,255),2)][num2hex(dna.GetUIValueRange(DNA_UI_BODY_MARK_G,255),2)][num2hex(dna.GetUIValueRange(DNA_UI_BODY_MARK_B,255),2)]"
+		marking_colours["tail"] = "#[num2hex(dna.GetUIValueRange(DNA_UI_TAIL_MARK_R,255),2)][num2hex(dna.GetUIValueRange(DNA_UI_TAIL_MARK_G,255),2)][num2hex(dna.GetUIValueRange(DNA_UI_TAIL_MARK_B,255),2)]"
```
These look like they should be using the `rgb` proc.

Krausus commented on code/game/dna/dna2_helpers.dm:
```diff
@@ -181,16 +181,43 @@
+			H.m_styles = list2params(marking_styles)
+		//Tail Markings
+		var/tail_marks = dna.GetUIValueRange(DNA_UI_TAIL_MARK_STYLE,marking_styles_list.len)
+		if((0 < tail_marks) && (tail_marks <= number_tail_marks))
```
The coding guidelines say to not use `0 < variable`; this should be `variable > 0`.

Krausus commented on code/game/dna/genes/vg_powers.dm:
```diff
@@ -46,82 +46,119 @@
 	if(new_eyes)
-		M.r_eyes = hex2num(copytext(new_eyes, 2, 4))
-		M.g_eyes = hex2num(copytext(new_eyes, 4, 6))
-		M.b_eyes = hex2num(copytext(new_eyes, 6, 8))
```
This removes the ability to change your eye color with morph.

KasparoVy replied:
No it doesn't, change_eye_color does the exact same thing. It was doing it twice. Tested in game, verified that the eye colour vars were changed for both the mob and the eye organ specifically in varedit.

Krausus replied:
The color you pick is stored in `new_eyes`, but then `new_eyes` isn't used for anything. If something actually changes your eye color, it isn't this.

Krausus commented on code/game/dna/genes/vg_powers.dm:
```diff
@@ -46,82 +46,119 @@
+	var/datum/sprite_accessory/hair_style = hair_styles_list[head_organ.h_style]
+	if(hair_style.secondary_theme && !hair_style.no_sec_colour)
+		new_hair = input("Please select secondary hair color.", "Character Generation", rgb(head_organ.r_hair_sec, head_organ.g_hair_sec, head_organ.b_hair_sec)) as null|color
+		M.change_hair_color(hex2num(copytext(new_hair, 2, 4)), hex2num(copytext(new_hair, 4, 6)), hex2num(copytext(new_hair, 6, 8)), 1)
```
Doesn't sanity check that `new_hair` is not null.

Krausus commented on code/game/dna/genes/vg_powers.dm:
```diff
@@ -46,82 +46,119 @@
+	var/datum/sprite_accessory/facial_hair_style = facial_hair_styles_list[head_organ.f_style]
+	if(facial_hair_style.secondary_theme && !facial_hair_style.no_sec_colour)
+		new_facial = input("Please select secondary facial hair color.", "Character Generation", rgb(head_organ.r_facial_sec, head_organ.g_facial_sec, head_organ.b_facial_sec)) as null|color
+		M.change_facial_hair_color(hex2num(copytext(new_facial, 2, 4)), hex2num(copytext(new_facial, 4, 6)), hex2num(copytext(new_facial, 6, 8)), 1)
```
Doesn't sanity check that `new_facial` is not null.

Krausus commented on code/game/dna/genes/vg_powers.dm:
```diff
@@ -46,82 +46,119 @@
+
+		var/list/marking_colours = params2list(M.m_colours)
+		marking_colours["head"] = sanitize_hexcolor(marking_colours["head"])
+		var/new_marking_colour = input("Please select head marking colour.", "Character Generation", rgb(hex2num(copytext(marking_colours["head"], 2, 4)), hex2num(copytext(marking_colours["head"], 4, 6)), hex2num(copytext(marking_colours["head"], 6, 8)))) as null|color
```
This takes a hex color, converts it to separate R/G/B values, then converts it back to a hex color. Why.

Krausus commented on code/game/dna/genes/vg_powers.dm:
```diff
@@ -46,82 +46,119 @@
+
+		var/list/marking_colours = params2list(M.m_colours)
+		marking_colours["body"] = sanitize_hexcolor(marking_colours["body"])
+		var/new_marking_colour = input("Please select body marking colour.", "Character Generation", rgb(hex2num(copytext(marking_colours["body"], 2, 4)), hex2num(copytext(marking_colours["body"], 4, 6)), hex2num(copytext(marking_colours["body"], 6, 8)))) as null|color
```
hexcolor->R/G/B->hexcolor again

Krausus commented on code/game/dna/genes/vg_powers.dm:
```diff
@@ -46,82 +46,119 @@
+
+		var/list/marking_colours = params2list(M.m_colours)
+		marking_colours["tail"] = sanitize_hexcolor(marking_colours["tail"])
+		var/new_marking_colour = input("Please select tail marking colour.", "Character Generation", rgb(hex2num(copytext(marking_colours["tail"], 2, 4)), hex2num(copytext(marking_colours["tail"], 4, 6)), hex2num(copytext(marking_colours["tail"], 6, 8)))) as null|color
```
hexcolor->R/G/B->hexcolor *again*.

Krausus commented on code/modules/client/preference/preferences.dm:
```diff
@@ -1383,9 +1467,86 @@ var/global/list/special_role_times = list( //minimum age (in days) for accounts
+						var/input = "Choose the colour of your your character's head markings:"
+						var/list/marking_colours = params2list(m_colours)
+						marking_colours["head"] = sanitize_hexcolor(marking_colours["head"])
+						var/new_markings = input(user, input, "Character Preference", rgb(hex2num(copytext(marking_colours["head"], 2, 4)), hex2num(copytext(marking_colours["head"], 4, 6)), hex2num(copytext(marking_colours["head"], 6, 8)))) as color|null
```
hexcolor->R/G/B->hexcolor

Krausus commented on code/modules/client/preference/preferences.dm:
```diff
@@ -1383,9 +1467,86 @@ var/global/list/special_role_times = list( //minimum age (in days) for accounts
+						var/input = "Choose the colour of your your character's body markings:"
+						var/list/marking_colours = params2list(m_colours)
+						marking_colours["body"] = sanitize_hexcolor(marking_colours["body"])
+						var/new_markings = input(user, input, "Character Preference", rgb(hex2num(copytext(marking_colours["body"], 2, 4)), hex2num(copytext(marking_colours["body"], 4, 6)), hex2num(copytext(marking_colours["body"], 6, 8)))) as color|null
```
hexcolor->R/G/B->hexcolor

Krausus commented on code/modules/client/preference/preferences.dm:
```diff
@@ -1383,9 +1467,86 @@ var/global/list/special_role_times = list( //minimum age (in days) for accounts
+						var/input = "Choose the colour of your your character's tail markings:"
+						var/list/marking_colours = params2list(m_colours)
+						marking_colours["tail"] = sanitize_hexcolor(marking_colours["tail"])
+						var/new_markings = input(user, input, "Character Preference", rgb(hex2num(copytext(marking_colours["tail"], 2, 4)), hex2num(copytext(marking_colours["tail"], 4, 6)), hex2num(copytext(marking_colours["tail"], 6, 8)))) as color|null
```
hexcolor->R/G/B->hexcolor

Krausus commented on code/modules/client/preference/preferences.dm:
```diff
@@ -1956,11 +2145,9 @@ var/global/list/special_role_times = list( //minimum age (in days) for accounts
-		character.g_markings = g_markings
-		character.b_markings = b_markings
-		character.m_style = m_style
+	if((character.species.bodyflags & HAS_HEAD_MARKINGS) || (character.species.bodyflags & HAS_BODY_MARKINGS) || (character.species.bodyflags & HAS_TAIL_MARKINGS))
```
You should keep `HAS_MARKINGS` as a combo of these 3 flags, instead of checking each individually.

Krausus commented on code/modules/customitems/item_defines.dm:
This should use `rgb`.

Krausus commented on code/modules/mob/living/carbon/human/human.dm:
```diff
@@ -1687,14 +1693,15 @@
-			var/optic_colour = input(src, "Select optic colour", rgb(r_markings, g_markings, b_markings)) as color|null
+			var/list/marking_colours = params2list(m_colours)
+			marking_colours["head"] = sanitize_hexcolor(marking_colours["head"])
+			var/optic_colour = input(src, "Select optic colour", rgb(hex2num(copytext(marking_colours["head"], 2, 4)), hex2num(copytext(marking_colours["head"], 4, 6)), hex2num(copytext(marking_colours["head"], 6, 8)))) as color|null
```
hexcolor->R/G/B->hexcolor

Krausus commented on code/modules/mob/living/carbon/human/update_icons.dm:
```diff
@@ -360,29 +360,38 @@ var/global/list/damage_icon_parts = list()
+			var/icon/b_marking_s = new/icon("icon" = body_marking_style.icon, "icon_state" = "[body_marking_style.icon_state]_s")
+			if(body_marking_style.do_colouration)
+				marking_colours["body"] = sanitize_hexcolor(marking_colours["body"])
+				b_marking_s.Blend(rgb(hex2num(copytext(marking_colours["body"], 2, 4)), hex2num(copytext(marking_colours["body"], 4, 6)), hex2num(copytext(marking_colours["body"], 6, 8))), ICON_ADD)
```
hexcolor->R/G/B->hexcolor

Krausus commented on code/modules/mob/living/carbon/human/update_icons.dm:
```diff
@@ -360,29 +360,38 @@ var/global/list/damage_icon_parts = list()
+			var/icon/h_marking_s = new/icon("icon" = head_marking_style.icon, "icon_state" = "[head_marking_style.icon_state]_s")
+			if(head_marking_style.do_colouration)
+				marking_colours["head"] = sanitize_hexcolor(marking_colours["head"])
+				h_marking_s.Blend(rgb(hex2num(copytext(marking_colours["head"], 2, 4)), hex2num(copytext(marking_colours["head"], 4, 6)), hex2num(copytext(marking_colours["head"], 6, 8))), ICON_ADD)
```
hexcolor->R/G/B->hexcolor

Krausus commented on code/modules/mob/living/carbon/human/update_icons.dm:
```diff
@@ -1121,142 +1143,140 @@ var/global/list/damage_icon_parts = list()
+		marking_colours["tail"] = sanitize_hexcolor(marking_colours["tail"])
+		tail_marking_style = marking_styles_list[tail_marking]
+		tail_marking_icon = new/icon("icon" = tail_marking_style.icon, "icon_state" = "[tail_marking_style.icon_state]_s")
+		tail_marking_icon.Blend(rgb(hex2num(copytext(marking_colours["tail"], 2, 4)), hex2num(copytext(marking_colours["tail"], 4, 6)), hex2num(copytext(marking_colours["tail"], 6, 8))), ICON_ADD)
```
hexcolor->R/G/B->hexcolor

Krausus commented on code/modules/mob/living/carbon/human/update_icons.dm:
```diff
@@ -1121,142 +1143,140 @@ var/global/list/damage_icon_parts = list()
+		marking_colours["tail"] = sanitize_hexcolor(marking_colours["tail"])
+		tail_marking_style = marking_styles_list[tail_marking]
+		tail_marking_icon = new/icon("icon" = tail_marking_style.icon, "icon_state" = "[tail_marking_style.icon_state]w_s")
+		tail_marking_icon.Blend(rgb(hex2num(copytext(marking_colours["tail"], 2, 4)), hex2num(copytext(marking_colours["tail"], 4, 6)), hex2num(copytext(marking_colours["tail"], 6, 8))), ICON_ADD)
```
hexcolor->R/G/B->hexcolor

I'm going to stop leaving comments on these, because this is just getting silly. You should be able to just search for `rgb(hex2num` and find every instance of this.

Krausus commented:
Two things I didn't bother commenting on because they're used all over the place:

`hex2num(copytext(hexcolor, 2, 4)), hex2num(copytext(hexcolor, 4, 6)), hex2num(copytext(hexcolor, 6, 8))`

The readability of these is *awful.* They should, at the very least, be replaced with helper procs - `color2R`, `color2G`, etc. It would be better for procs to support using RGB hex colors directly, so conversion could happen in one place (if it even needs to) instead of every time the proc is called.

```
var/list/marking_styles = params2list(H.m_styles)
H.m_styles = list2params(marking_styles)
```

`m_styles` should absolutely not be a params string. The only time you don't convert it is when you're storing it in the database, but that should be the only time you *do* convert it; the rest of the time, it should be a list that can be used directly. Looking at the diffs, you call `params2list` over **60** times throughout this code. You only actually need to call it *twice.*

KasparoVy commented:
Thanks @Krausus for proofreading this, didn't get any of this input on the last iteration of the PR and I'll be going through all you've pointed out right now.

Krausus commented on code/game/dna/genes/vg_powers.dm:
```diff
@@ -44,84 +44,114 @@
 	if(new_eyes)
-		M.r_eyes = hex2num(copytext(new_eyes, 2, 4))
-		M.g_eyes = hex2num(copytext(new_eyes, 4, 6))
-		M.b_eyes = hex2num(copytext(new_eyes, 6, 8))
```
As mentioned previously, removing these removes any use of the `new_eyes` var. What gets input here isn't used whatsoever.

KasparoVy replied:
Woops, I guess I was thinking of this as I would the eye colour changing up in nanoUI human_appearance. The line below has now been changed to get it working again.

Krausus commented on code/game/dna/dna2.dm:
```diff
@@ -154,9 +162,12 @@ var/global/list/bad_blocks[0]
-		character.m_style = "None"
-	var/marks	= marking_styles_list.Find(character.m_style)
+	if(!character.m_styles)
+		character.m_styles = initial(character.m_styles)
```
Due to how BYOND initializes lists, initial doesn't work with them. You'll probably want to put their initial value in a `#define` and use that anywhere you initialize it.

Krausus commented on code/modules/client/preference/preferences_mysql.dm:
```diff
@@ -300,6 +313,8 @@
 	if(!organ_data) src.organ_data = list()
 	if(!rlimb_data) src.rlimb_data = list()
+	if(!m_colours) m_colours = initial(m_colours)
+	if(!m_styles) m_styles = initial(m_styles)
```
These initials also won't work.

Krausus commented on code/modules/mob/living/carbon/human/human.dm:
```diff
@@ -1559,30 +1559,36 @@
 		H.b_headacc = 0
 
+	m_styles = initial(m_styles) //Wipes out markings, setting them all to "None".
+	m_colours = initial(m_styles) //Defaults colour to #00000 for all markings.
```
More initials.

KasparoVy commented:
Sorry for the late response, been out all day. Thanks again, never knew that about lists. Made the change to morph eye colour handling and set up the defines, just testing before I push the commit.

KasparoVy commented:
Done. However, I seem to have uncovered another issue, one that I have verified to exist on the current master branch of our codebase.

The set_species proc [here](https://github.com/ParadiseSS13/Paradise/blob/e944f176d61c346fdf760c9491e87f770926a725/code/modules/mob/living/carbon/human/human.dm#L1484) only seems to correctly set the (facial) hair style/colour and head accessory style/colour when creating a new mob (on the master branch), but when you call set_species("Skrell") on a Human mob, for example, instead of getting the Skrell Male tentacles hair style like they're supposed to they keep the same hairstyle as they did previously (although it won't render on their sprite) because the hair style, colour and whatever vars that were supposed to be set in the set_species proc **weren't.** Only body_accessory seems to get correctly nulled out-- although languages and stuff are correctly adjusted, too. They aren't bugged. This issue persists in this PR as well as extends to the setting of marking styles/colours.

In example: Spawn a Skrell from the game panel. That is where set_species is fully operational. Then spawn a human/human and set_species("Skrell") them to see the above described bug.

Strange! Got me scratching my head and I can't seem to figure it out. I think the spawn proc might help here, but I've no experience with it.

KasparoVy commented:
Picture added to showcase the mask alt_head refitting.

Fox-McCloud commented:
Requesting additional review by @Krausus

Fox-McCloud commented:
@KasparoVy 

will have to update this to account for the serialization changes.

KasparoVy commented:
Sure thing, getting on that now

KasparoVy commented:
Sorted out my git problems and the conflicts shortly afterward

KasparoVy commented on code/modules/client/preference/preferences.dm:
```diff
@@ -1850,20 +1985,34 @@ var/global/list/special_role_times = list( //minimum age (in days) for accounts
-
 	//Head-specific
 	var/obj/item/organ/external/head/H = character.get_organ("head")
+	var/obj/item/organ/internal/eyes/eyes_organ = character.get_int_organ(/obj/item/organ/internal/eyes)
```
This and the following 4 lines are subject to change as per the status of https://github.com/ParadiseSS13/Paradise/pull/5495

Fox-McCloud commented:
Merge conflicts again because of Crazy's fix.

KasparoVy commented:
Thanks again, should be all sorted now

QuinnAggeler commented:
How difficult do you think it might be to add limb-based markings?

KasparoVy commented:
Are you talking about limb markings for the Drask, so they could change the colour of their arm spines?
That wouldn't be difficult at all, it'll just take a bit of copy-pasting and time. Although, I could be wrong. I'll look into it.

However, I think a more complete limb-based marking system (working the icon handling into the cached icon generation of update_body) would be reasonably complex and it's something I've been saving as the 'final evolution' of the marking system.

There's no way I'm adding it in this PR as it would delay it at least until mid September, probably later

What I had in mind for that was to make it so choosable markings are comprised of various components, the icon states for which following the nomenclature [species.name]_[body_part.name]_markingname etc etc

QuinnAggeler commented:
Aye, for the Drask. Then they will finally have proper customization

Fox-McCloud commented:
@KasparoVy 

Please don't add any more to this PR beyond what is necessary to get this working.

KasparoVy commented:
@Fox-McCloud It's done, I'll leave the rest for another PR.

KasparoVy commented:
@Fox-McCloud It's done (pending review), I'll leave the rest for another time. No worries

Krausus commented on code/modules/client/preference/preferences_mysql.dm:
```diff
@@ -332,23 +349,28 @@
-												markings_red='[r_markings]',
-												markings_green='[g_markings]',
-												markings_blue='[b_markings]',
+												marking_colours='[markingcolourslist]',
```
Any time you put text into a query, it needs to be passed through `sql_sanitize_text()`, even if it should never contain characters that would break the query string.

Krausus commented on code/modules/client/preference/preferences_mysql.dm:
```diff
@@ -332,23 +349,28 @@
 												hair_style_name='[sql_sanitize_text(h_style)]',
 												facial_style_name='[sql_sanitize_text(f_style)]',
-												marking_style_name='[sql_sanitize_text(m_style)]',
+												marking_styles='[markingstyleslist]',
```
`sql_sanitize_text()`

Krausus commented on code/modules/client/preference/preferences_mysql.dm:
```diff
@@ -418,11 +442,13 @@
+											'[r_facial_sec]', '[g_facial_sec]', '[b_facial_sec]',
 											'[s_tone]', '[r_skin]', '[g_skin]', '[b_skin]',
-											'[r_markings]', '[g_markings]', '[b_markings]',
+											'[markingcolourslist]',
```
`sql_sanitize_text()`

Krausus commented on code/modules/client/preference/preferences_mysql.dm:
```diff
@@ -418,11 +442,13 @@
+											'[markingcolourslist]',
 											'[r_headacc]', '[g_headacc]', '[b_headacc]',
-											'[sql_sanitize_text(h_style)]', '[sql_sanitize_text(f_style)]', '[sql_sanitize_text(m_style)]', '[sql_sanitize_text(ha_style)]',
+											'[sql_sanitize_text(h_style)]', '[sql_sanitize_text(f_style)]', '[markingstyleslist]', '[sql_sanitize_text(ha_style)]', '[sql_sanitize_text(alt_head)]',
```
`sql_sanitize_text()`

Krausus commented on code/modules/mob/new_player/sprite_accessories.dm:
```diff
@@ -1878,70 +1962,270 @@
+
+/datum/sprite_accessory/body_markings/points_fade_belly_vulp
+	name = "Vulpkanin Points and Belly"
+	species_allowed = list("Vulpkanin")
```
`list("Species Name")` is copy-pasted a *ton* all across this file. It should really be on a subtype, so instead of `body_markings/points_fade_belly_vulp`, you have `body_markings/vulp/points_fade_belly`, with `body_markings/vulp` being the one that sets `species_allowed`. You'll want to make sure the sprite accessory list builder skips convenience subtypes, of course.

Also, that would make it easy to split species-specific sprites into separate DMIs, so we don't end up with everything crammed into a single massive DMI.

KasparoVy replied:
Sounds like a good idea to me, thanks for the detail. I'd like to leave that for another PR that squares it all away at once. Think there'd be any performance advantage that would come along with such a change, too?

Krausus commented:
Everything else looks fine. Our appearance handling is too much of a mess for me to take the time to try and reason about how well all of this works, though; if everything works fine in testing, then it's probably fine.

We really need a massive refactor, though. I'm pretty sure we've got three different places with slightly different code handling appearance modification, and that's terrible. I just hope nobody decides to cram even more code into--

> However, I think a more complete limb-based marking system (working the icon handling into the cached icon generation of update_body) would be reasonably complex and it's something I've been saving as the 'final evolution' of the marking system.

:frowning:

Krausus commented:
Upon closer examination, `sql_sanitize_text()` is actually awful and will ruin the params strings. Hmm.

Krausus commented:
So, turns out the proper proc is `sanitizeSQL()`. `sql_sanitize_text()` is being completely removed by #5500, so that will cause merge conflicts. Sorry~

KasparoVy commented:
No worries, thanks again for the help Krausus I'll get back on it and make the changes :+1:

KasparoVy commented:
Hella late but something came up that caused a decent delay. Conflicts should be resolved again, added sanitization to the marking SQL handling and tested it by loading a character in game, saving a randomized character, loading that character again, checking the entries in my database and then loading in as a civ with the characters.

KasparoVy commented:
Turns out I did miss a spot, but I caught it. Forgot to sanitizeSQL markingcolourslist on 468 in the first pass but it's been changed and tested. Everything seems in order.

KasparoVy commented:
Found out what the deal with that obscure species changing bug was. For whatever reason if you call set_species without the 'delay icon update' argument the result is a mob with a different species but all the same cosmetic attributes as the old species (meaning the resets and the changing of attributes was ignored during the execution of the proc), but that's sorted now.

Fixed marking resetting, tidied up a bit and added some more markings.

@HugoLuman It's just a body marking (meaning I didn't work in another marking location for this) but it produces the same visual effect. This character is a result of clicking the randomize button twice, what luck!

![draskspinemarking](https://cloud.githubusercontent.com/assets/12377767/17888572/95cee2a0-68f9-11e6-9dee-bc1672dfadc8.png)

KasparoVy commented:
Found out what the deal with that obscure species changing bug was. For whatever reason if you call set_species without the 'delay icon update' argument the result is a mob with a different species but all the same cosmetic attributes as the old species (meaning the resets and the changing of attributes was ignored during the execution of the proc), but that's sorted now.

Fixed marking resetting, tidied up a bit and added some more markings.

I also found out what the issue was with fatness. Because of how update_body worked prior to this PR, despite a mob becoming fat/slim their body parts wouldn't be updated to reflect that, thus only their uniforms would get a fat sprite and not their bodies. Now, because I've implemented an override in update_body that prompts re-caching of body part icons, if a mob's overeatduration is over 500 it will actually look fat (if it's a species that can be fat and has fat sprites) else it will look skinny.

Tested by setting overeatduration to 666 when naked as a human, became fat. Tested again by setting overeatduration to 0 when naked as a human, became thin. Did the same with a black jumpsuit on. Jumpsuit got the fat/skinny sprite. Took off the jumpsuit and noticed I was fat/skinny.

@HugoLuman It's just a body marking (meaning I didn't work in another marking location for this) but it produces the same visual effect and can be accessed in preferences by checking out the new 'body markings' section that will pop up for Drask. This character is a result of clicking the randomize button twice, what luck!

_Drask are finally able to colour their arm-spines._
![draskspinemarking](https://cloud.githubusercontent.com/assets/12377767/17888572/95cee2a0-68f9-11e6-9dee-bc1672dfadc8.png)

_Showcasing the Vulpkanin markings I just added._
![dobermanvulpmarkings](https://cloud.githubusercontent.com/assets/12377767/17888751/74890b24-68fa-11e6-8c8b-e68787f99613.png)![shibevulpmarkings](https://cloud.githubusercontent.com/assets/12377767/17888752/748a3dd2-68fa-11e6-9e6d-43c606cb7a65.png)

Changelog updated and pictures added.

KasparoVy commented:
Found out what the deal with that obscure species changing bug was. For whatever reason if you call set_species without the 'delay icon update' argument the result is a mob with a different species but all the same cosmetic attributes as the old species (meaning the resets and the changing of attributes was ignored during the execution of the proc), but that's sorted now.

Fixed marking resetting, tidied up a bit and added some more markings.

I also found out what the issue was with fatness. Because of how update_body worked prior to this PR, despite a mob becoming fat/slim their body parts wouldn't be updated to reflect that, thus only their uniforms would get a fat sprite and not their bodies. Now, because I've implemented an override in update_body that prompts re-caching of body part icons, if a mob's overeatduration is over 500 it will actually look fat (if it's a species that can be fat and has fat sprites) else it will look skinny. #5421 was halfway there.

Tested by setting overeatduration to 666 when naked as a human, became fat. Tested again by setting overeatduration to 0 when naked as a human, became thin. Did the same with a black jumpsuit on. Jumpsuit got the fat/skinny sprite. Took off the jumpsuit and noticed I was fat/skinny.
![bigboi](https://cloud.githubusercontent.com/assets/12377767/17889615/12027558-68ff-11e6-8807-da49e46733f8.PNG)

Tweaked my fix to the mutation bit. Didn't test it before, assumed it worked. Tested it, found out it didn't. Small tweak sorted it out. I tested it by dealing a mob 5 cloneloss over and over and over and over and over, witnessing each bodypart grotesquely mutate.
![ohgod](https://cloud.githubusercontent.com/assets/12377767/17892761/e5e3d87e-6910-11e6-92b8-ab4d55e757b5.png)

@HugoLuman It's just a body marking (meaning I didn't work in another marking location for this) but it produces the same visual effect and can be accessed in preferences by checking out the new 'body markings' section that will pop up for Drask. This character is a result of clicking the randomize button twice, what luck!

_Drask are finally able to colour their arm-spines._
![draskspinemarking](https://cloud.githubusercontent.com/assets/12377767/17888572/95cee2a0-68f9-11e6-9dee-bc1672dfadc8.png)

_Showcasing the Vulpkanin markings I just added._
![dobermanvulpmarkings](https://cloud.githubusercontent.com/assets/12377767/17888751/74890b24-68fa-11e6-8c8b-e68787f99613.png)![shibevulpmarkings](https://cloud.githubusercontent.com/assets/12377767/17888752/748a3dd2-68fa-11e6-9e6d-43c606cb7a65.png)

Changelog updated and pictures added.

KasparoVy commented:
DNM pending a bugfix and the repathing of the stuff in sprite_accessories.dm

KasparoVy commented:
Alrighty, that takes care of the runtimes and sprite_accessories.dm.

KasparoVy commented:
Alrighty, that takes care of the runtimes and sprite_accessories.dm. I tested out all the lists by going through the preferences screen, the nanomirrors and CMA and everything's as it should be. No blank list entries or nothing.

Added randomization for all the new stuff + the stuff I added a while back but forgot to do and a couple other small fixes. Shouldn't be any conflicts anymore, either.

KasparoVy commented:
Alrighty, that takes care of the runtimes and sprite_accessories.dm. I tested out all the lists by going through the preferences screen, the nanomirrors and CMA and everything's as it should be. No blank list entries or nothing.

Added randomization (after doing it for when you change species in preferences and forgetting to do it for the actual randomization button) for all the new stuff + the stuff I added a while back but forgot to do and a couple other small fixes. Shouldn't be any conflicts anymore, either.

KasparoVy commented:
Thanks for putting DNM on so fast! I had a hunch I missed a spot and an issue was where I suspected it'd be. Mostly oversight.

Alrighty, that takes care of the runtimes and sprite_accessories.dm. I tested out all the lists by going through the preferences screen, the nanomirrors and CMA and everything's as it should be. No blank list entries polluting the selection screens or nothing.

Added randomization (after doing it for when you change species in preferences and forgetting to do it for the actual randomization button) for all the new stuff + the stuff I added a while back but forgot to do and a couple other small fixes. Shouldn't be any conflicts anymore, either.
Can go ahead and remove those two labels if you'd like

KasparoVy commented:
Thanks for putting DNM on so fast! I had a hunch I missed a spot and an issue was where I suspected it'd be. Mostly oversight.

Alrighty, that takes care of the runtimes and sprite_accessories.dm. I tested out all the lists by going through the preferences screen, the nanomirrors and CMA and everything's as it should be. No blank list entries polluting the selection screens or nothing.

Added randomization (after doing it for when you change species in preferences and forgetting to do it for the actual randomization button) for all the new stuff + the stuff I added a while back but forgot to do and a couple other small fixes.
Corrected some oversight where I forgot to make it such that when a dead body gets skeletonized the head accessory, markings and alt-heads are wiped.

Shouldn't be any conflicts anymore, either.
Can go ahead and remove those two labels if you'd like

KasparoVy commented:
Thanks for putting DNM on so fast! I had a hunch I missed a spot and an issue was where I suspected it'd be. Mostly oversight.

Alrighty, that takes care of the runtimes and sprite_accessories.dm. I tested out all the lists by going through the preferences screen, the nanomirrors and CMA and everything's as it should be. No blank list entries polluting the selection screens or nothing.

Added randomization (after doing it for when you change species in preferences and forgetting to do it for the actual randomization button) for all the new stuff + the stuff I added a while back but forgot to do and a couple other small fixes.
Corrected some oversight where I forgot to make it such that when a dead body gets skeletonized the head accessory, markings and alt-heads are wiped.

Shouldn't be any conflicts anymore, either.
Can go ahead and remove those two labels if you'd like

KasparoVy commented:
More conflicts, gonna see what's up

HugoLuman commented:
https://cloud.githubusercontent.com/assets/12377767/17888572/95cee2a0-68f9-11e6-9dee-bc1672dfadc8.png
I LOVE YOU

Spacemanspark commented:
RUN KASP

KasparoVy commented:
Oh God, what's happening?
